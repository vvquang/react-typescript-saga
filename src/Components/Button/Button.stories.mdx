import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks'
import { action } from '@storybook/addon-actions'
import Button from '.'
import Space from 'stories/components/styled-system/Space.mdx'
import Styles from 'stories/styles.mdx'

<Meta title="Story/Component/Button" component={Button} />

<Styles />

# Button

With `MDX` we can define a story for `Button` right in the middle of our
Markdown documentation.

export const Template = (args) => <Button {...args} />

<Canvas>
  <Story
    name="Large"
    args={{ size: 'large' }}
    onClick={action('Large button clicked')}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="Medium"
    args={{ size: 'medium' }}
    onClick={action('Medium button clicked')}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="Disable"
    args={{ disabled: true }}
    onClick={action('Disable button clicked')}
  >
    {Template.bind({})}
  </Story>
</Canvas>

<div class="subheading">Props</div>

<Space />
