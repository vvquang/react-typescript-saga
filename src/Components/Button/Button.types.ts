import {
  SpaceProps,
  LayoutProps,
  FlexBasisProps,
  ButtonStyleProps,
} from 'styled-system'

interface ThemeButton {
  color?: string
  kind?: string
}

interface ButtonProps {
  theme?: ThemeButton
  label?: string
  size?: 'small' | 'medium' | 'large'
}

export declare type BaseButtonProps = React.ButtonHTMLAttributes<
  HTMLButtonElement
> &
  ButtonProps &
  SpaceProps &
  LayoutProps &
  FlexBasisProps &
  ButtonStyleProps

export interface IButtonProps extends BaseButtonProps {
  children?: React.ReactNode
  props?: BaseButtonProps
  ref?: any
}
