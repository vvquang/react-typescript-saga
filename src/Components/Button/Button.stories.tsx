import React from 'react'
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from '@storybook/react/types-6-0'

// import {
//   Title,
//   Subtitle,
//   Description,
//   Primary,
//   ArgsTable,
//   Stories,
//   PRIMARY_STORY,
// } from '@storybook/addon-docs/blocks'

import { Button } from '.'
import { IButtonProps } from './Button.types'

export default {
  title: 'Story/Component/Button',
  component: Button,
  argTypes: {
    size: {
      control: {
        type: 'select',
        options: ['large', 'medium'],
      },
    },
  },
  // parameters: {
  //   docs: {
  //     page: () => (
  //       <>
  //         <Title />
  //         <Subtitle />
  //         <Description />
  //         <Primary />
  //         <ArgsTable story={PRIMARY_STORY} />
  //         <Stories />
  //       </>
  //     ),
  //   },
  // },
} as Meta

const Template: Story<IButtonProps> = (args) => (
  <Button {...args}>Button story</Button>
)

export const Large = Template.bind({})
Large.args = {
  size: 'large',
}

export const Small = Template.bind({})
Small.args = {
  size: 'small',
  label: 'Small button',
}

export const Disable = Template.bind({})
Disable.args = {
  disabled: true,
}
