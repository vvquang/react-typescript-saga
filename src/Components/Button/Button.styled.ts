import styled, { css } from 'styled-components'
import { prop, ifProp, switchProp } from 'styled-tools'

import { space, layout, flexbox, buttonStyle } from 'styled-system'
import { primaryPalette /*, COLOR  */ } from 'Theme/colors'
import { BaseButtonProps } from './Button.types'

export const AppButton = styled.button<BaseButtonProps>`
  ${space};
  ${layout};
  ${flexbox};
  ${buttonStyle};

  color: ${prop('theme.colors.white', '#fff')};
  font-size: ${ifProp(
    { size: 'large' },
    prop('theme.sizes.lg', '20px'),
    prop('theme.sizes.md', '14px')
  )};
  background-color: ${prop('theme.colors.black', '#000')};

  ${switchProp('kind', {
    dark: css`
      background-color: ${prop('theme.colors.blue', 'blue')};
      border: 1px solid ${prop('theme.colors.blue', 'blue')};
    `,
    darker: css`
      background-color: ${prop('theme.colors.mediumblue', 'mediumblue')};
      border: 1px solid ${prop('theme.colors.mediumblue', 'mediumblue')};
    `,
    darkest: css`
      background-color: ${prop('theme.colors.darkblue', 'darkblue')};
      border: 1px solid ${prop('theme.colors.darkblue', 'darkblue')};
    `,
  })}

  ${ifProp(
    'disabled',
    css`
      background-color: ${prop('theme.colors.gray', '#999')};
      border-color: ${prop('theme.colors.gray', '#999')};
      pointer-events: none;
    `
  )}
`

export const theme = {
  palette: {
    primary: primaryPalette,
    secondary: ['#c2185b', '#e91e63', '#f06292', '#f8bbd0'],
  },
}
