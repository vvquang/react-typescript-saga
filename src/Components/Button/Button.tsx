import React, { FC, forwardRef } from 'react'
import { ThemeProvider } from 'styled-components'
import { AppButton, theme } from './Button.styled'
import { IButtonProps } from './Button.types'

const Button: FC<IButtonProps> = forwardRef(
  (props, ref, children = 'Button') => {
    return (
      <ThemeProvider theme={theme}>
        <AppButton
          ref={ref}
          size="medium"
          variant="normal"
          theme={{ kind: 'dark' }}
          {...props}
        >
          {children}
        </AppButton>
      </ThemeProvider>
    )
  }
)

Button.displayName = 'Button'
export default Button
