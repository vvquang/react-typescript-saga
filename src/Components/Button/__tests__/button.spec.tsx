import React from 'react'
import Button from '..'
import { mount } from 'enzyme'

// NOTE:: Using mount from enzyme to test atoms that we can test any css changed in children dom
// Ref: https://blog.halolabs.io/testing-react-components-with-jest-enzyme-5d1dd4ddccc4
describe('Button', () => {
  it('primary large btn', () => {
    const primaryLargeBtn = mount(<Button size="large">Primary</Button>)
    expect(primaryLargeBtn).toMatchSnapshot()

    const expectStyle: { [key: string]: string } = {
      'font-size': '20px',
      'background-color': '#000',
    }

    for (const style in expectStyle) {
      if (expectStyle[style]) {
        expect(primaryLargeBtn).toHaveStyleRule(style, expectStyle[style])
      }
    }
  })

  it('disable btn', () => {
    const outlineLargeBtn = mount(<Button disabled>Disable button</Button>)
    expect(outlineLargeBtn).toMatchSnapshot()

    const expectStyle: { [key: string]: string } = {
      'background-color': '#999',
      'border-color': '#999',
      'pointer-events': 'none',
    }

    for (const style in expectStyle) {
      if (expectStyle[style]) {
        expect(outlineLargeBtn).toHaveStyleRule(style, expectStyle[style])
      }
    }
  })
})
