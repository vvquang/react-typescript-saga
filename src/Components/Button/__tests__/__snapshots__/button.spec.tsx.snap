// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button disable btn 1`] = `
.c0 {
  width: medium;
  height: medium;
  color: #fff;
  font-size: 14px;
  background-color: #000;
  background-color: #999;
  border-color: #999;
  pointer-events: none;
}

<Button
  disabled={true}
>
  <Component
    theme={
      Object {
        "palette": Object {
          "primary": Object {
            "contrastText": "#FFF",
            "dark": "#cf2557",
            "light": "#fc688e",
            "main": "#F62F5E",
          },
          "secondary": Array [
            "#c2185b",
            "#e91e63",
            "#f06292",
            "#f8bbd0",
          ],
        },
      }
    }
  >
    <styled.button
      disabled={true}
      size="medium"
      theme={
        Object {
          "kind": "dark",
        }
      }
      variant="normal"
    >
      <button
        className="c0"
        disabled={true}
        size="medium"
      >
        Button
      </button>
    </styled.button>
  </Component>
</Button>
`;

exports[`Button primary large btn 1`] = `
.c0 {
  width: large;
  height: large;
  color: #fff;
  font-size: 20px;
  background-color: #000;
}

<Button
  size="large"
>
  <Component
    theme={
      Object {
        "palette": Object {
          "primary": Object {
            "contrastText": "#FFF",
            "dark": "#cf2557",
            "light": "#fc688e",
            "main": "#F62F5E",
          },
          "secondary": Array [
            "#c2185b",
            "#e91e63",
            "#f06292",
            "#f8bbd0",
          ],
        },
      }
    }
  >
    <styled.button
      size="large"
      theme={
        Object {
          "kind": "dark",
        }
      }
      variant="normal"
    >
      <button
        className="c0"
        size="large"
      >
        Button
      </button>
    </styled.button>
  </Component>
</Button>
`;
